cocoa/src/toga_cocoa/window.py

```diff
@@ -40,6 +40,60 @@ def windowDidResize_(self, notification) -> None:
             # Set the window to the new size
             self.interface.content.refresh()
 
+    @objc_method
+    def windowDidBecomeMain_(self, notification):
+        self.impl.interface.on_gain_focus()
+
+    @objc_method
+    def windowDidResignMain_(self, notification):
+        self.impl.interface.on_lose_focus()
+
+    @objc_method
+    def windowDidBecomeKey_(self, notification):
+        if bool(self.impl.native.isVisible) and not self.impl._is_previously_shown:
+            self.impl._is_previously_shown = True
+            self.impl.interface.on_show()
+
+    @objc_method
+    def windowDidMiniaturize_(self, notification):  # pragma: no cover
+        if not bool(self.impl.native.isVisible) and self.impl._is_previously_shown:
+            self.impl._is_previously_shown = False
+            self.impl.interface.on_hide()
+
+    @objc_method
+    def windowDidDeminiaturize_(self, notification):  # pragma: no cover
+        if bool(self.impl.native.isVisible) and not self.impl._is_previously_shown:
+            self.impl._is_previously_shown = True
+            self.impl.interface.on_show()
+
+    @objc_method
+    def windowDidEnterFullScreen_(self, notification):  # pragma: no cover
+        if bool(self.impl.native.isVisible) and not self.impl._is_previously_shown:
+            self.impl._is_previously_shown = True
+            self.impl.interface.on_show()
+
+    @objc_method
+    def windowDidExitFullScreen_(self, notification):  # pragma: no cover
+        if bool(self.impl.native.isVisible) and not self.impl._is_previously_shown:
+            self.impl._is_previously_shown = True
+            self.impl.interface.on_show()
+
+    # when the user clicks the zoom button to unzoom a window
+    @objc_method
+    def windowWillUseStandardFrame_defaultFrame_(
+        self, window, defaultFrame
+    ):  # pragma: no cover
+        if bool(self.impl.native.isVisible) and not self.impl._is_previously_shown:
+            self.impl._is_previously_shown = True
+            self.impl.interface.on_show()
+
+    # when the user clicks the zoom button to zoom a window
+    @objc_method
+    def windowShouldZoom_toFrame_(self, window, toFrame):  # pragma: no cover
+        if bool(self.impl.native.isVisible) and not self.impl._is_previously_shown:
+            self.impl._is_previously_shown = True
+            self.impl.interface.on_show()
```

Comment:
I had added `self.impl._is_previously_shown` because minimizing a window is same as losing focus to the native OS. 

But from the API viewpoint, minimizing a window should mean that it is not visible to the user and also it has lost focus. 

`isVisible` would be `True` even when the window is minimized. But as per the use case of the API, the window contents are practically not visible to the user and the app can pause updating its contents. 

These two cases are necessary as the window can be maximized(making the window visible to the user) from a minimized state(not visible to the user).

Reply:
Right - but that just reinforces that the issue is the absence of a window state API. You've added an internal state variable to work around the fact that you can't ask a window if it's minimized. The right approach here isn't the add a workaround - you fix the underlying problem.

However, more generally, my comment still stands. This particular endpoint is invoked when a macOS window goes "full screen" as a result of hitting the "maximise" button. That *requires* that the window is on the screen, at some size, with its title bar visible. You can't click a maximise button that isn't there. So the window *must* already be in a shown state. Similar, if it's in a maximized state, it *must* be on the screen, with a title bar visible. 

So why are these two handlers required? What's the edge case where a window goes from a "not shown" state to a "shown" state *as a result of being maximized*.

Reply:
>What's the edge case where a window goes from a "not shown" state to a "shown" state as a result of being maximized.

The one edge case I had in mind was if the window was made full screen(Visible to user) from a minimized state(Not Visible to user) through API. In this case, the window doesn't needs to be in a maximized state for it become full screen.

I agree a window state API would be better. I will start working on it as soon as the other dependent PRs are completed.